onnxruntime/core/providers/cuda/fpgeneric.cu

```diff
@@ -64,11 +64,22 @@ __global__ void CopyVectorBFloat16(const onnxruntime::BFloat16* x, int incx, onn
 
 }  // namespace
 
+dim3 cublasTransposeHelperDimGrid(int m, int n) {
+  return {(n + TRANS_TILE_DIM - 1) / TRANS_TILE_DIM, (m + TRANS_TILE_DIM - 1) / TRANS_TILE_DIM, 1};
+}
+
+// cublasTransposeHelper can only be used if it won't overflow the 65536 grid dimension size
+__host__ bool CanUse_cublasTransposeHelper_MLFloat16(int m, int n) {
+  dim3 dimGrid = cublasTransposeHelperDimGrid(m, n);
+  return dimGrid.x < 65536 && dimGrid.y < 65536;
```

Comment:
Is there a way to programmatically get the max blocks per grid device property and check against that ? Like device.maxBlocksPerGrid ? Or is it always bound to be 65535 ?

Reply:
There likely is, the table here shows that it varies for x, but not for y. x goes up to 2^31-1 starting with compute 3.0. y is 65535 max for all versions, and z is 3 for all versions:

https://en.wikipedia.org/wiki/CUDA#Version_features_and_specifications

You could probably modify it to take advantage of that, but it's not a few hour fix.

Reply:
The condition can be `dimGrid.y < 65536`. The x has larger limit. Did you test a case of dimGrid.x > 65536 && dimGrid.y < 65536?

https://en.wikipedia.org/wiki/Thread_block_(CUDA_programming)#:~:text=Blocks%20can%20be%20organized%20into,y%20and%20z%20dimensions%20respectively.

Reply:
Fixed. I tried adding a test case for the dimGrid.x but couldn't decipher the shape/perm requirements to make it happen. It kept making dimGrid.x very low with what I tried.